Userland/Libraries/LibGfx/Path.cpp

```diff
@@ -365,4 +365,166 @@ void Path::add_path(Path const& other)
     invalidate_split_lines();
 }
 
+template<typename T>
+struct RoundTrip {
+    RoundTrip(ReadonlySpan<T> span)
+        : m_span(span)
+    {
+    }
+
+    size_t size() const
+    {
+        return m_span.size() * 2 - 1;
+    }
+
+    T const& operator[](size_t index) const
+    {
+        // Follow the path:
+        if (index < m_span.size())
+            return m_span[index];
+        // Then in reverse:
+        if (index < size())
+            return m_span[size() - index - 1];
+        // Then wrap around again:
+        return m_span[index - size() + 1];
+    }
+
+private:
+    ReadonlySpan<T> m_span;
+};
+
+Path Path::stroke_to_fill(float thickness) const
+{
+    // Note: This convolves a polygon with the path using the algorithm described
+    // in https://keithp.com/~keithp/talks/cairo2003.pdf (3.1 Stroking Splines via Convolution)
+
+    auto& lines = split_lines();
+    if (lines.is_empty())
+        return Path {};
+
+    // Paths can be disconnected, which a pain to deal with, so split it up.
+    Vector<Vector<FloatPoint>> segments;
+    segments.append({ lines.first().a() });
+    for (auto& line : lines) {
+        if (line.a() == segments.last().last()) {
+            segments.last().append(line.b());
+        } else {
+            segments.append({ line.a(), line.b() });
+        }
+    }
+
+    // Note: This is the same as the tolerance from bezier curve splitting.
+    constexpr auto flatness = 0.015f;
+    auto pen_vertex_count = max(
+        static_cast<int>(ceilf(AK::Pi<float> / acosf(1 - (2 * flatness) / thickness))), 4);
+    if (pen_vertex_count % 2 == 1)
+        pen_vertex_count += 1;
+
+    Vector<FloatPoint, 128> pen_vertices;
+    pen_vertices.ensure_capacity(pen_vertex_count);
+
+    // Generate vertices for the pen (going counterclockwise). The pen does not necessarily need
+    // to be a circle (or an approximation of one), but other shapes are untested.
+    float theta = 0;
+    float theta_delta = (AK::Pi<float> * 2) / pen_vertex_count;
+    for (int i = 0; i < pen_vertex_count; i++) {
+        float sin_theta;
+        float cos_theta;
+        AK::sincos(theta, sin_theta, cos_theta);
+        pen_vertices.unchecked_append({ cos_theta * thickness / 2, sin_theta * thickness / 2 });
+        theta -= theta_delta;
+    }
+
+    auto wrapping_index = [](auto& vertices, auto index) {
+        return vertices[(index + vertices.size()) % vertices.size()];
+    };
+
+    auto angle_between = [](auto p1, auto p2) {
+        auto delta = p2 - p1;
+        return atan2f(delta.y(), delta.x());
+    };
+
+    struct ActiveRange {
+        float start;
+        float end;
+
+        bool in_range(float angle) const
+        {
+            // Note: Since active ranges go counterclockwise start > end unless we wrap around at 180 degrees
+            return ((angle <= start && angle >= end)
+                || (start <= end && angle <= start)
+                || (start <= end && angle >= end));
```

Comment:
These last two lines probably don't hold when `start == end`, check should maybe be `start < end` instead?

Reply:
I think so, I don't think there should be any case where `start == end` (for a circular pen)